File: ruby/test/receipt_base64_fast_path_test.rb
# frozen_string_literal: true

require_relative "helper"

# The strict fast path in front of the receipt-data base64 decoder.
#
# Receipt.decode_base64 tries `unpack1("m0")` first and falls back to the
# tolerant rule on anything it declines. That is only a speed-up if it can
# never change an answer: every string the fast path accepts must be one the
# tolerant path accepts, with the same bytes, and every rejection must still
# carry the tolerant path's reason and message. A client whose receipt was
# refused yesterday must be refused today for the same stated reason, and a
# receipt that decoded must decode to the same DER.
class ReceiptBase64FastPathTest < Minitest::Test
  APRV = ApplePurchaseReceiptVerifier
  RECEIPT = APRV::Receipt

  SEED = 20_260_922
  RANDOM_INPUTS = 20_000

  # Strings where a strict decoder and the tolerant rule are most likely to
  # disagree: padding in every wrong count and place, whitespace, the other
  # alphabet, non-ASCII, and the lengths base64 cannot have.
  EDGE_CASES = [
    "", " ", "\n", "=", "==", "===", "====", "A", "AA", "AAA", "AAAA", "A===",
    "AA=", "AA==", "AA===", "AAA=", "AAA==", "AAAA=", "AAAA==", "AAAA===",
    "AAAA====", "AAAAA===", "AB==", "AAB=", "QQ==QQ==", "AA==AA==", "AAA=AAA=",
    "AAAA\n", " AAAA", "AA\r\nAA", "AA-_", "AA+/", "A+_A", "AAAA!", "éééé",
    "AA\x00A", "////", "++++", "QUJD".encode(Encoding::UTF_16LE)
  ].freeze

  WHITESPACE = " \t\r\n"
  POOL = [*"A".."Z", *"a".."z", *"0".."9", "+", "/", "-", "_", "=", " ", "\t", "\r", "\n", "!", "é"].freeze

  # One input, drawn so that each branch of the decoder gets thousands of
  # hits: clean canonical base64, the variants only the tolerant path
  # accepts, and near misses that both must reject.
  def random_input(random)
    canonical = [random.bytes(random.rand(40))].pack("m0")
    case random.rand(6)
    when 0 then canonical
    when 1
      url = canonical.tr("+/", "-_")
      random.rand < 0.5 ? url.delete("=") : url
    when 2
      chars = canonical.chars
      random.rand(1..3).times { chars.insert(random.rand(chars.size + 1), WHITESPACE[random.rand(4)]) }
      chars.join
    when 3 then canonical.delete("=") + ("=" * random.rand(5))
    when 4 then near_miss(random, canonical.chars)
    else Array.new(random.rand(13)) { POOL.sample(random: random) }.join
    end
  end

  def near_miss(random, chars)
    position = random.rand(chars.size + 1)
    operation = random.rand(3)
    if operation.zero? || chars.empty?
      chars.insert(position, POOL.sample(random: random))
    elsif operation == 1
      chars[[position, chars.size - 1].min] = POOL.sample(random: random)
    else
      chars.delete_at([position, chars.size - 1].min)
    end
    chars.join
  end

  def inputs
    random = Random.new(SEED)
    EDGE_CASES + Array.new(RANDOM_INPUTS) { random_input(random) }
  end

  def outcome(text)
    [:ok, yield(text)]
  rescue APRV::VerificationError => e
    [:error, e.reason, e.message]
  end

  def tolerant(text)
    outcome(text) { |t| RECEIPT.decode_base64_tolerant(t) }
  end

  # The inputs on which `fast` in front of the tolerant path answers
  # differently from the tolerant path alone.
  def mismatches(texts, &fast)
    texts.reject do |text|
      combined = outcome(text) { |t| fast.call(t) || RECEIPT.decode_base64_tolerant(t) }
      combined == tolerant(text)
    end
  end

  def test_the_fast_path_never_changes_an_answer
    branches = { fast: 0, tolerant_accept: 0, reject: 0 }
    inputs.each do |text|
      expected = tolerant(text)
      assert_equal expected, outcome(text) { |t| RECEIPT.decode_base64(t) }, text.inspect
      if RECEIPT.decode_base64_strict(text).nil?
        branches[expected.first == :ok ? :tolerant_accept : :reject] += 1
      else
        # Accepted by the fast path, so the tolerant path must accept it
        # too; the equality above already compared the bytes.
        assert_equal :ok, expected.first, text.inspect
        branches[:fast] += 1
      end
    end
    # A differential test that never reached one of the branches proves
    # nothing about it.
    branches.each { |branch, hits| assert_operator hits, :>, 1000, "#{branch}: #{branches}" }
  end

  # The comparison above can fail: a decoder that skips characters outside
  # the alphabet instead of refusing them accepts strings the tolerant rule
  # rejects, and the same inputs expose it.
  def test_a_permissive_fast_path_is_caught
    refute_empty(mismatches(inputs) { |text| text.empty? ? nil : text.unpack1("m") })
  end

  # `unpack1("m0")` decodes "" to "", which the tolerant rule rejects as an
  # empty receipt. The guard is what keeps the fast path a subset; it is the
  # only one needed on the Rubies this was measured on.
  def test_the_empty_guard_is_what_makes_the_decoder_a_subset
    unguarded = mismatches(EDGE_CASES) do |text|
      text.unpack1("m0")
    rescue ArgumentError
      nil
    end
    assert_equal [""], unguarded
    assert_empty(mismatches(EDGE_CASES) { |text| RECEIPT.decode_base64_strict(text) })
  end
end

File: node/test/receipt-base64-fast-path.test.js
// decodeReceiptDataString decodes canonical standard base64 through
// Buffer.from before trying the full receipt-data decode. Buffer.from
// accepts anything, so the only thing keeping the verdicts unchanged is the
// gate in front of it. This compares the decoder against the full decode
// alone (receiptBase64DecodeStrict plus the error it maps to) on a seeded
// corpus built to sit on both sides of that gate: equal bytes when both
// accept, the same reason and message when both reject, and never one
// accepting what the other rejects.
import test from 'node:test';
import assert from 'node:assert/strict';
import { receiptBase64DecodeStrict } from '../dist/bytes.js';
import { decodeReceiptDataString } from '../dist/receipt.js';

/** The full decode with the fast path taken out: what the answers must equal. */
function fullDecode(text) {
  const decoded = receiptBase64DecodeStrict(text);
  if (decoded === null) {
    return { error: 'INVALID_RECEIPT_FORMAT: receipt-data is not valid base64' };
  }
  return { bytes: Buffer.from(decoded).toString('hex') };
}

function withFastPath(text) {
  try {
    return { bytes: decodeReceiptDataString(text).toString('hex') };
  } catch (error) {
    return { error: `${error.reason}: ${error.message.slice(error.reason.length + 2)}` };
  }
}

// Mirrors the gate in src/receipt.ts, only to count which branch each input
// took; the comparison itself does not depend on it.
const takesFastPath = (text) =>
  text.length !== 0 && text.length % 4 === 0 && /^[A-Za-z0-9+/]*={0,2}$/.test(text);

// mulberry32: small, seeded, and the same sequence on every run.
function rng(seed) {
  let state = seed >>> 0;
  return () => {
    state = (state + 0x6d2b79f5) >>> 0;
    let t = state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
const EXTRAS = ['-', '_', '=', ' ', '\n', '\r', '\t', '*', '.', 'é', '\u0000', 'Ａ'];

function corpus(count, seed) {
  const random = rng(seed);
  const int = (n) => Math.floor(random() * n);
  const pick = (list) => list[int(list.length)];
  const inputs = [];
  for (let i = 0; i < count; i++) {
    const bytes = Buffer.from(Array.from({ length: int(48) }, () => int(256)));
    let text = bytes.toString('base64');
    switch (int(12)) {
      case 0:
      case 1:
      case 2:
        break; // canonical, the fast path's whole domain
      case 3: // flip the unused low bits of the last data character
        text = text.replace(/[A-Za-z0-9+/](?==*$)/, () => pick([...ALPHABET]));
        break;
      case 4:
        text = text.replace(/=+$/, '');
        break;
      case 5:
        text += '='.repeat(1 + int(3));
        break;
      case 6:
        text = bytes.toString('base64url');
        break;
      case 7: {
        const at = int(text.length + 1);
        text = text.slice(0, at) + pick(EXTRAS) + text.slice(at);
        break;
      }
      case 8:
        text = text.slice(0, Math.max(0, text.length - 1 - int(3)));
        break;
      case 9: // random characters from both alphabets, whitespace and padding
        text = Array.from({ length: int(24) }, () =>
          random() < 0.8 ? ALPHABET[int(64)] : pick(EXTRAS),
        ).join('');
        break;
      case 10: // line-wrapped the way Foundation does it
        text = text.replace(/(.{4})/g, (run) => (random() < 0.3 ? `${run}\r\n` : run));
        break;
      default: {
        const at = int(text.length);
        text = text.slice(0, at) + pick([...ALPHABET, ...EXTRAS]) + text.slice(at + 1);
      }
    }
    inputs.push(text);
  }
  return inputs;
}

const EDGE_CASES = [
  '',
  ' ',
  '=',
  '==',
  '===',
  '====',
  'A',
  'A=',
  'A==',
  'A===',
  'AA',
  'AA=',
  'AA==',
  'AA===',
  'AB==',
  'AAA',
  'AAA=',
  'AAB=',
  'AAAA',
  'AAAA=',
  'AAAA====',
  '=AAA',
  'A=AA',
  'AA=A',
  '+/+/',
  '-_-_',
  '+/-_',
  '//==',
  'AAAA\n',
  ' AAAA',
  'AAéA',
  'ＡAAA',
];

test('the fast path answers what the full decode answers, input for input', () => {
  const inputs = [...EDGE_CASES, ...corpus(24_000, 0x5eed1e55)];
  const branches = { fast: 0, fullAccept: 0, reject: 0 };
  let ungatedWouldDiffer = 0;
  for (const text of inputs) {
    const expected = fullDecode(text);
    assert.deepEqual(withFastPath(text), expected, JSON.stringify(text));
    if (takesFastPath(text)) {
      assert.ok(expected.bytes !== undefined, `the gate let through ${JSON.stringify(text)}`);
      branches.fast += 1;
    } else if (expected.bytes !== undefined) {
      branches.fullAccept += 1;
    } else {
      branches.reject += 1;
    }
    if (
      expected.bytes === undefined ||
      Buffer.from(text, 'base64').toString('hex') !== expected.bytes
    ) {
      ungatedWouldDiffer += 1;
    }
  }
  // Each branch carried a real share of the corpus, so the equality above
  // was checked on all three.
  for (const [branch, count] of Object.entries(branches)) {
    assert.ok(count > 1000, `${branch} saw only ${count} inputs`);
  }
  // The corpus has teeth: Buffer.from without the gate would have answered
  // differently on thousands of these inputs.
  assert.ok(ungatedWouldDiffer > 1000, `an ungated fast path differs on ${ungatedWouldDiffer}`);
});
